### PR TITLE
Testing Ruby with Rspec (Tic Tac Toe)

### PR DESCRIPTION
**Pull request by dansantander**
This project complies with the following requirements:

- Add specs for all public methods you created in Tic Tac Toe project.
- Use existing repositories for Tic Tac Toe: add specs and open new Pull Requests.
- Wrap specs for each method in describe block.
- Instead of using before_each to define a variable for the test cases, use let.
- Use one it block for each expected behaviour. Do not test multiple aspects in one it block.
- Use your common sense: test positive and negative scenarios.

### CODE REVIEW

File: spec/tictactoe_spec.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+require_relative '../lib/board.rb'
+require_relative '../lib/gamelogic.rb'
+require 'colorize'
+
+describe Board do
+  let(:board) { Board.new }
+  X = 'X'.green
+  O = 'O'.blue
+
+  describe '#win' do
```
Comment (bolah2009) [inline spec/tictactoe_spec.rb:12]: Could you test for negative cases too?
Currently, the test is only testing for when `(board.win).to eql(1)` however the method return `1`, `2` and `false`.

Commenting out the below code passes still passes the test.

```ruby
  def win
    win_arr = [[0, 1, 2], [3, 4, 5], [6, 7, 8],
               [0, 3, 6], [1, 4, 7], [2, 5, 8],
               [0, 4, 8], [2, 4, 6]]

    win_arr.each do |block|
      return 1 if block.all? { |cell| @arr[cell] == X }
      return 2 if block.all? { |cell| @arr[cell] == O }
    end
    # false
  end
```

File: spec/tictactoe_spec.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  describe '#valid user name input' do
+    it "checks if player's name is not an alphabet string" do
+      expect(game.valid_name_input('1')).to eql(false)
+    end
+
+    it "checks if player's name is not empty" do
+      expect(game.valid_name_input('')).to eql(false)
+    end
+  end
```
Comment (bolah2009) [inline spec/tictactoe_spec.rb]: Could we also test for cases where `valid_name_input` is true? Commenting out the code below still passes the test.


```ruby
  def valid_name_input(input)
    if input[/[^a-zA-Z]/] || input == ''
      false
    # else
    #   true
    end
  end
```

File: spec/tictactoe_spec.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      expect(game.winner(p1, p2, board)).to eql(p1.name.green)
+    end
+
+    it "returns which player wins the match in it's due color" do
+      board.arr = [O, O, O, X, O, X, X, X, O] # player 2 wins
+      expect(game.winner(p1, p2, board)).to eql(p2.name.blue)
+    end
+  end
+
+  describe '#score' do
```
Comment (bolah2009) [inline spec/tictactoe_spec.rb:149]: See the previous comments on testing negative cases.

File: spec/tictactoe_spec.rb [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    it "checks if player's name is not an alphabet string" do
+      expect(game.valid_name_input('1')).to eql(false)
+    end
+
+    it "checks if player's name is not empty" do
+      expect(game.valid_name_input('')).to eql(false)
+    end
+  end
+
+  describe '#winner' do
```
Comment (bolah2009) [inline spec/tictactoe_spec.rb:132]: See the previous comments on testing negative cases.

**Review by bolah2009 — CHANGES_REQUESTED** [review]
Good job with the testing... 👍🏾

However, I would recommend testing navigate cases for public methods. See comment for more details.

>  A bit of common sense: check if specs cover positive and negative scenarios ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/commonsense.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/)) 
Example: there is a spec for correct usage of method \[e.g. `it "creates object if all required attributes are passed"`\] and also there is spec for wrong usage of method \[e.g. `it "does not create object if required attributes are missing"`\].

#### :recycle:  Make these changes and submit another code review request. 👍🏾

#### Feel free to reach me on slack if you having issues implementing these features. :relaxed: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

**Review by mosesogwo — APPROVED** [review]
**You've done well making the recommended changes:+1:. Project is approved :heavy_check_mark:**

**_Good luck in your next project:stars:_**

### Status: Approved.

**Pull request merged**